File: 백준/Silver/1780. 종이의 개수/종이의 개수.swift
import Foundation

// nM1, n0, nP1: Int (mutable)
// description: An integer that stores the number of papers with different values
var nM1: Int = 0
var n0: Int = 0
var nP1: Int = 0

// paper: [[Int]] (mutable)
// description: Each value of points in given paper
// range: -1, 0, 1
// length: N*N
var paper = [[Int]]()

// recursion: func(Int, Int, Int)
// description: A recursive function that checks if all elements in a square 
// submatrix are the same, and if not, divides it into 9 zones and recursively processes each
// x: Int (start row index)
// y: Int (start column index)
// size: Int (length of square submatrix)
func recursion(_ x: Int, _ y: Int, _ size: Int) {
    // first: First value in current submatrix
    let first: Int = paper[x][y]

    // same: Bool (flag)
    // description: Flag to indicate if all values in the current submatrix are the same
    var same: Bool = true

    // Check if all values in submatrix are the same as first
    for i in x..<x + size {
        for j in y..<y + size {
            if paper[i][j] != first {
                same = false
                break
            }
        }
        if !same { break }
    }

    // If all values are the same, increment corresponding counter
    if same {
        if first == -1 {
            nM1 += 1
        } else if first == 0 {
            n0 += 1
        } else {
            nP1 += 1
        }
        return
    }

    // newSize: Int
    // description: Size of each divided zone (1/3 of current size)
    let newSize: Int = size / 3

    // Recursive call: divide into 9 zones
    for dx in 0..<3 {
        for dy in 0..<3 {
            // x + dx * newSize, y + dy * newSize: Starting point of subzone
            recursion(x + dx * newSize, y + dy * newSize, newSize)
        }
    }

    return
}

// N: Int
// description: size of given paper
// range: 1 ≤ N ≤ 3^7 (3^k)
if let inputOfN: String = readLine(), let N: Int = Int(inputOfN) {
    // Append each line to list "paper" 
    for _ in 0..<N {
        if let line: String = readLine() {
            paper.append( line.split(separator: " ").compactMap { Int($0) } )
        }
    }

    // Call function
    recursion(0, 0, N)

    // Print answer
    print(nM1)
    print(n0)
    print(nP1)
}

File: 백준/Silver/1780. 종이의 개수/종이의 개수.py
import sys
	
input = sys.stdin.readline
print = sys.stdout.write

nM1 = 0
n0 = 0
nP1 = 0

# N: int
# description: size of given paper
# range: 1 ≤ N ≤ 3^7 (3^k)
N = int(input())

# paper: list[list[int]]
# description: Each value of points in given paper
# range: -1, 0, 1
# length: N*N
paper = [list(map(int, input().split())) for _ in range(N)]

# isSingleValue: def(list[list[int]]) -> bool
# description: A function that returns boolean based on wheter the values 
# in a given 2-Dimensional list are all the same value
def isSingleValue(v: list[list[int]]) -> bool:
    # firstValue: First value of list "v"
    firstValue = v[0][0]

    # Traveling the list v, return False if val is different value from the "firstValue"
    for row in v:
        for val in row:
            if val != firstValue:
                return False
    else:
        # Return True when iteration completes search to the end.
        return True

# recursion: def(list[list[int]])
# description: A recursive function that divided by 9 zones 
# a given 2-Dimensional list and update number of each values
def recursion(v: list[list[int]]):
    global nM1, n0, nP1
    # n: Number of rows in list v
    n = len(v)

    # Return first value when no longer divisible or is a single value
    if (n == 1 or isSingleValue(v) == True):
        val = v[0][0]
        if (val == -1):
            nM1 += 1
        elif (val == 0):
            n0 += 1
        else:
            nP1 += 1
        return

    tri = n//3

    top = v[:tri]
    mid = v[tri: tri*2]
    bot = v[tri*2:]

    # Recurvise call divided by 9 zones 
    recursion([row[:tri] for row in top])
    recursion([row[tri: tri*2] for row in top])
    recursion([row[tri*2:] for row in top])
    recursion([row[:tri] for row in mid])
    recursion([row[tri: tri*2] for row in mid])
    recursion([row[tri*2:] for row in mid])
    recursion([row[:tri] for row in bot])
    recursion([row[tri: tri*2] for row in bot])
    recursion([row[tri*2:] for row in bot])

    return

# Call function
recursion(paper)

# Print answer
print(f"{nM1}\n{n0}\n{nP1}\n")